lib/srv/desktop/rdp/rdpclient/src/lib.rs

```diff
@@ -92,62 +94,118 @@ impl IronRDPClient {
-/// Go, the "current thread" can be thought of as whichever goroutine the exported rdp function is called from. Because the client might
-/// be being used by multiple goroutines concurrently, it is up to the programmer to consider any synchronization mechanisms that might
-/// need to be implemented as features are added to the Client going forward.
-pub struct Client {
-    iron_rdp_client: IronRDPClient,
-    tokio_rt: tokio::runtime::Runtime,
-    go_ref: usize,
-}
-
+/// After calling [`Client::new()`], the caller is responsible for calling [`Client::drop()`].
+/// Letting the Client go out of scope without calling [`Client::drop()`] will result in a memory
+/// leak.
+///
+/// Using the Client after calling [`Client::drop()`] will result in undefined behavior.
+///
+/// Client implements [`Deref`] to [`ClientInternal`], which means that it can be
+/// used directly as a &ClientInternal in Rust code.
+#[repr(C)]
+pub struct Client(*const ClientInternal);
 impl Client {
+    /// Creates a new [`Client`]
     fn new(
         iron_rdp_client: IronRDPClient,
         go_ref: usize,
         tokio_rt: tokio::runtime::Runtime,
-    ) -> *mut Self {
-        Box::into_raw(Box::new(Self {
+    ) -> Self {
+        Self(Box::into_raw(Box::new(ClientInternal::new(
             iron_rdp_client,
-            tokio_rt,
             go_ref,
-        }))
+            tokio_rt,
+        ))))
     }
 
-    unsafe fn from_raw(ptr: *mut Self) -> Result<&'static mut Client, CGOErrCode> {
-        match ptr.as_ref() {
-            None => {
-                error!("invalid Rust client pointer");
-                Err(CGOErrCode::ErrCodeClientPtr)
-            }
-            Some(_) => Ok(Box::leak(Box::from_raw(ptr))),
-        }
+    /// # Safety
+    ///
+    /// This function is unsafe because improper use may lead to
+    /// memory problems. For example, a double-free may occur if the
+    /// function is called twice on the same Client.
+    ///
+    /// While Rust's memory semantics make this impossible in pure Rust,
+    /// it is possible to call this function twice from Go.
+    unsafe fn drop(self) {
+        drop(Box::from_raw(self.0 as *mut ClientInternal));
+    }
+
+    fn new_null() -> Self {
+        Self(ptr::null())
     }
+}
+
+impl Deref for Client {
+    type Target = ClientInternal;
+
+    #[inline]
+    fn deref(&self) -> &Self::Target {
+        unsafe { &*self.0 }
+    }
+}
+
+/// # Safety
+///
+/// Client is a pointer to a ClientInternal, which is Send
+/// per assert_impl_all!(Client: Send, Sync);
+unsafe impl Send for Client {}
+/// # Safety
+///
+/// Client is a pointer to a ClientInternal, which is Sync per
+/// assert_impl_all!(Client: Send, Sync);
+/// Client only implements Deref to ClientInernal, meaning that
+/// only immutable references of ClientInternal (themselves Sync)
+/// can be used.
+unsafe impl Sync for Client {}
```

Comment:
`T` being `Send + Sync` does not imply that `*const T` is `Send + Sync`.  I'm not knowledgeable enough to come up with all the ways this breaks things, but an obvious one is that these trait impls would allow a reference to be captured by a divergent thread, which is not something that is safe for `*const T` (yes, I know we're already doing that by passing a `*const T` to go, but we do that under the understanding that go is taking _ownership_ of that pointer and therefore it is "safe" in that its go's responsibility to ensure its safe usage... the rust side of things ensures safe usage by following rust's rules, and this impl breaks our ability to enforce those).

Reply:
> `T` being `Send + Sync` does not imply that `*const T` is `Send + Sync`

Granted my justification in the comments makes it sound like I'm saying that, however that's not what I'm trying to argue.

What you're pointing out is somewhat of a tautology given that ["raw pointers are neither Send nor Sync (because they have no safety guards)."](https://doc.rust-lang.org/nomicon/send-and-sync.html). A counter to your point is that just because `*const T` is never _automatically_ `Send + Sync` does not mean that `*const T` is not `Send + Sync`. 

 What I'm trying to claim is something like in the example given in this section of the Rustonomicon (linked above):

> Rc and UnsafeCell are very fundamentally not thread-safe: they enable unsynchronized shared mutable state. However raw pointers are, strictly speaking, marked as thread-unsafe as more of a lint. Doing anything useful with a raw pointer requires dereferencing it, which is already unsafe. In that sense, one could argue that it would be "fine" for them to be marked as thread safe.
>
> However it's important that they aren't thread-safe to prevent types that contain them from being automatically marked as thread-safe. These types have non-trivial untracked ownership, and it's unlikely that their author was necessarily thinking hard about thread safety. In the case of Rc, we have a nice example of a type that contains a *mut that is definitely not thread-safe.
>
> Types that aren't automatically derived can simply implement them if desired:
>
> ```rust
> struct MyBox(*mut u8);
> 
> unsafe impl Send for MyBox {}
> unsafe impl Sync for MyBox {}
> ```

My argument is, for this type that has "non-trivial untracked ownership", we actually _were_ thinking hard about thread safety, and this type has it. We know this because the only (non-unsafe) thing you can do with it is `Deref`, and the `Deref` gives you a `&ClientInternal`. So this `Client` is essentially a "proxy" to an `&ClientInternal`. Therefore we can argue:

1. ["`&T` is `Send` if and only if `T` is `Sync`"](https://doc.rust-lang.org/std/marker/trait.Sync.html) --> `&ClientInternal` is `Send` because `ClientInternal` is `Sync` --> `Client` is a proxy to `&ClientInternal` --> **`Client` is `Send`**
2. [`&T` and `&mut T` are `Sync` if and only if `T` is `Sync`](https://doc.rust-lang.org/std/marker/trait.Sync.html) --> `&ClientInternal` is `Sync` because `ClientInternal` is `Sync` --> `Client` is a proxy to `&ClientInternal` --> **`Client` is `Sync`**

Reply:
Sorry I wasn't clear enough, I didn't just mean that `*const T` might not be `Send` + `Sync`, so this isn't.  I brought up the example of thread capture because this type `in particular` cannot be safely passed to a divergent thread, and these trait implementations say that it is.

Let me give a more specific example:

 Say that we moved to a shared executor instead of one executor per client and we started spawning all work into that shared executor.  If we were just casting `*const T` to `&T` and doing all of our work on `&T` rust would immediately complain, because rust would know that spawning these references into a shared executor might cause a reference to be captured by a divergent thread.  No problem, we'd find a workaround.  The important thing is that the compiler could easily see and prevent the potential access of invalid memory, because we'd accurately described the constraints of the type to the compiler.

With this wrapper, if we switched to spawning into a shared executor the compiler would look and say "you're moving an owned value that is `Send`, no problem!" and let us do it, but the reference is actually only valid until the end of the scope of function.  Then either, if we're very lucky, someone realizes that type `Client` is escaping into the executor, or we have UB/invalid memory access potentially happening inside the executor.

Point isn't wether or not the usage is correct *now*, point is that by implementing these traits we prevent the compiler from telling us if usage becomes incorrect in the future.

Reply:
Thanks for that example, that clears it up. I see that my reasoning was wrong because the compiler reasons differently about owned values than references when it comes moving them around. So you're right that this value actually isn't `Send`.

The problem is, the shared executor example you just gave actually _is_ our situation, in the sense that Go owns the `Client`, and the Go runtime is a shared executor which can arbitrarily spawn references to this `Client` on potentially divergent threads. That's the whole point of me trying to enforce `Send + Sync`ness in the first place.

So I think the goal here should be to make `Client` _actually be_ `Send + Sync`, because the underlying physical reality is that it's being used as if it is (even though the Rust compiler can't enforce that).

Reply:
> With this wrapper, if we switched to spawning into a shared executor the compiler would look and say "you're moving an owned value that is Send, no problem!" and let us do it, but the reference is actually only valid until the end of the scope of function. Then either, if we're very lucky, someone realizes that type Client is escaping into the executor, or we have UB/invalid memory access potentially happening inside the executor.

Actually, let's double check this now that i'm using `as_ref` like you suggested -- in this case, won't we now just get a panic rather than undefined behavior?

https://github.com/gravitational/teleport/blob/be966c37737eeda3b3288405dda51b8b4b4bc1e1/lib/srv/desktop/rdp/rdpclient/src/client.rs#L103-L119

Reply:
Ok, here's what I have: It's impossible to guarantee that Client is `Send + Sync` by reasoning about Rust's semantics because we're handing off ownership to Go (by giving it back as a raw pointer). There's no way we can guarantee Go doesn't do anything it wants with with that raw pointer including calling something on it after it was freed, etc.

This little "smart pointer" wrapper that I have over it is just confusing things, it's not really "smart" at all: it's giving an illusion of ownership, it's `deref` pretends to be safe but it actually isn't, it's ultimately a confusing abstraction. I'm just going to scrap this whole thing and go with what you suggested [here](https://github.com/gravitational/teleport/pull/29659#issuecomment-1652777995).

Reply:
You're 100% right that we're basically doing the same executor thing on the go side, but the go side gets to decide when to deallocate the memory, meaning that in the terms of rust's model, the go side has "ownership" of the memory.  It's taken responsibility for it. The goal on the rust side should be to represent that relationship in rust terms.  When one of the FFI functions (other than the `free`/`drop` one) is called, rust is "borrowing" the value from the go side, but the go side is maintaining ownership (the right to do whatever it wants with it including delete it once the function returns).  That means on the rust side we should represent it as a reference.  Since those functions might be called concurrently, they are shared references (`&T` rather than `&mut T`).

>  I'm starting to wonder if that's just impossible, though.

I'm fairly certain that you already fixed the problem by changing the concurrent methods from operating on `&mut T` to `&T` and asserting that the base client was `Send`/`Sync`.  Thats the fix, at least in terms of the rust side.  We know the client is safe to share across threads in theory, so long as rust's rules are correctly observed.  And now that all use of shared references are represented as `&T` (or `*const T` at the FFI layer), they are!

Now that the rust-side knows what kind of references they are, it can enforce access rules correctly.  So long as we ensure that the go side never calls the `free`/`drop` function until all other rust functions are no longer in-use, we are guaranteed that it is safe to call because the rules of rust promise that the references we shared, being `&T`, were constrained to the lifetime of the function and will not be accessed again.

Reply:
`*const T` vs `*mut T` is just a hint to the readers (and some confusing source of UB when casting, since `const_t_pointer as &mut T` is actually `const_t_pointer as &T as &mut T` which is never allowed, whereas `const_t_pointer as *mut T as &mut T` might work just fine if the pointer was converted from an exclusive reference).

Also, to be precise, `Send` is only required to drop the value from a OS thread that's not the initial one (as dropping is , all functions merely acting on a reference to the value only require `Sync`.